File: Source/Alternet.UI.Pal/AuiNotebook.cpp
#include "AuiNotebook.h"

namespace Alternet::UI
{
    class wxAuiNotebook2 : public wxAuiNotebook, public wxWidgetExtender
    {
    public:
        wxAuiNotebook2(wxWindow* parent,
            wxWindowID id = wxID_ANY,
            const wxPoint& pos = wxDefaultPosition,
            const wxSize& size = wxDefaultSize,
            long style = wxAUI_NB_DEFAULT_STYLE)
        {
            Create(parent, id, pos, size, style);
        }
    };

	AuiNotebook::AuiNotebook()
	{

	}

	AuiNotebook::~AuiNotebook()
	{
        if (IsWxWindowCreated())
        {
            auto window = GetWxWindow();
            if (window != nullptr)
            {
                window->Unbind(wxEVT_AUINOTEBOOK_PAGE_CLOSE,
                    &AuiNotebook::OnPageClose, this);
                window->Unbind(wxEVT_AUINOTEBOOK_PAGE_CLOSED,
                    &AuiNotebook::OnPageClosed, this);
                window->Unbind(wxEVT_AUINOTEBOOK_PAGE_CHANGED,
                    &AuiNotebook::OnPageChanged, this);
                window->Unbind(wxEVT_AUINOTEBOOK_PAGE_CHANGING,
                    &AuiNotebook::OnPageChanging, this);
                window->Unbind(wxEVT_AUINOTEBOOK_BUTTON,
                    &AuiNotebook::OnWindowListButton, this);
                window->Unbind(wxEVT_AUINOTEBOOK_BEGIN_DRAG,
                    &AuiNotebook::OnBeginDrag, this);
                window->Unbind(wxEVT_AUINOTEBOOK_END_DRAG,
                    &AuiNotebook::OnEndDrag, this);
                window->Unbind(wxEVT_AUINOTEBOOK_DRAG_MOTION,
                    &AuiNotebook::OnDragMotion, this);
                window->Unbind(wxEVT_AUINOTEBOOK_ALLOW_DND,
                    &AuiNotebook::OnAllowTabDrop, this);
                window->Unbind(wxEVT_AUINOTEBOOK_DRAG_DONE,
                    &AuiNotebook::OnDragDone, this);
                window->Unbind(wxEVT_AUINOTEBOOK_TAB_MIDDLE_DOWN,
                    &AuiNotebook::OnTabMiddleMouseDown, this);
                window->Unbind(wxEVT_AUINOTEBOOK_TAB_MIDDLE_UP,
                    &AuiNotebook::OnTabMiddleMouseUp, this);
                window->Unbind(wxEVT_AUINOTEBOOK_TAB_RIGHT_DOWN,
                    &AuiNotebook::OnTabRightMouseDown, this);
                window->Unbind(wxEVT_AUINOTEBOOK_TAB_RIGHT_UP,
                    &AuiNotebook::OnTabRightMouseUp, this);
                window->Unbind(wxEVT_AUINOTEBOOK_BG_DCLICK,
                    &AuiNotebook::OnBgDclickMouse, this);
            }
        }
    }

    void AuiNotebook::FromEventData(wxAuiNotebookEvent& event)
    {
        int _eventSelection = event.GetSelection();
        int _eventOldSelection = event.GetOldSelection();
    }

    void AuiNotebook::OnPageClose(wxAuiNotebookEvent& event)
    {
        FromEventData(event);
        RaiseEvent(AuiNotebookEvent::PageClose);
        event.Skip();
    }

    void AuiNotebook::OnPageClosed(wxAuiNotebookEvent& event)
    {
        FromEventData(event);
        RaiseEvent(AuiNotebookEvent::PageClosed);
        event.Skip();
    }

    void AuiNotebook::OnPageChanged(wxAuiNotebookEvent& event)
    {
        FromEventData(event);
        RaiseEvent(AuiNotebookEvent::PageChanged);
        event.Skip();
    }

    void AuiNotebook::OnPageChanging(wxAuiNotebookEvent& event)
    {
        FromEventData(event);
        RaiseEvent(AuiNotebookEvent::PageChanging);
        event.Skip();
    }

    void AuiNotebook::OnWindowListButton(wxAuiNotebookEvent& event)
    {
        FromEventData(event);
        RaiseEvent(AuiNotebookEvent::WindowListButton);
        event.Skip();
    }

    void AuiNotebook::OnBeginDrag(wxAuiNotebookEvent& event)
    {
        FromEventData(event);
        RaiseEvent(AuiNotebookEvent::BeginDrag);
        event.Skip();
    }

    void AuiNotebook::OnEndDrag(wxAuiNotebookEvent& event)
    {
        FromEventData(event);
        RaiseEvent(AuiNotebookEvent::EndDrag);
        event.Skip();
    }
        
    void AuiNotebook::OnDragMotion(wxAuiNotebookEvent& event)
    {
        FromEventData(event);
        RaiseEvent(AuiNotebookEvent::DragMotion);
        event.Skip();
    }

    void AuiNotebook::OnAllowTabDrop(wxAuiNotebookEvent& event)
    {
        FromEventData(event);
        RaiseEvent(AuiNotebookEvent::AllowTabDrop);
        event.Skip();
    }

    void AuiNotebook::OnDragDone(wxAuiNotebookEvent& event)
    {
        FromEventData(event);
        RaiseEvent(AuiNotebookEvent::DragDone);
        event.Skip();
    }
    
    void AuiNotebook::OnTabMiddleMouseDown(wxAuiNotebookEvent& event)
    {
        FromEventData(event);
        RaiseEvent(AuiNotebookEvent::TabMiddleMouseDown);
        event.Skip();
    }
        
    void AuiNotebook::OnTabMiddleMouseUp(wxAuiNotebookEvent& event)
    {
        FromEventData(event);
        RaiseEvent(AuiNotebookEvent::TabMiddleMouseUp);
        event.Skip();
    }

    void AuiNotebook::OnTabRightMouseDown(wxAuiNotebookEvent& event)
    {
        FromEventData(event);
        RaiseEvent(AuiNotebookEvent::TabRightMouseDown);
        event.Skip();
    }

    void AuiNotebook::OnTabRightMouseUp(wxAuiNotebookEvent& event)
    {
        FromEventData(event);
        RaiseEvent(AuiNotebookEvent::TabRightMouseUp);
        event.Skip();
    }
    
    void AuiNotebook::OnBgDclickMouse(wxAuiNotebookEvent& event)
    {
        FromEventData(event);
        RaiseEvent(AuiNotebookEvent::BgDclickMouse);
        event.Skip();
    }

    wxWindow* AuiNotebook::CreateWxWindowCore(wxWindow* parent)
    {
        long style = wxAUI_NB_DEFAULT_STYLE;

        auto window = new wxAuiNotebook2(parent,
            wxID_ANY,
            wxDefaultPosition,
            wxDefaultSize,
            style);

        window->Bind(wxEVT_AUINOTEBOOK_PAGE_CLOSE,
            &AuiNotebook::OnPageClose, this);
        window->Bind(wxEVT_AUINOTEBOOK_PAGE_CLOSED,
            &AuiNotebook::OnPageClosed, this);
        window->Bind(wxEVT_AUINOTEBOOK_PAGE_CHANGED,
            &AuiNotebook::OnPageChanged, this);
        window->Bind(wxEVT_AUINOTEBOOK_PAGE_CHANGING,
            &AuiNotebook::OnPageChanging, this);
        window->Bind(wxEVT_AUINOTEBOOK_BUTTON,
            &AuiNotebook::OnWindowListButton, this);
        window->Bind(wxEVT_AUINOTEBOOK_BEGIN_DRAG,
            &AuiNotebook::OnBeginDrag, this);
        window->Bind(wxEVT_AUINOTEBOOK_END_DRAG,
            &AuiNotebook::OnEndDrag, this);
        window->Bind(wxEVT_AUINOTEBOOK_DRAG_MOTION,
            &AuiNotebook::OnDragMotion, this);
        window->Bind(wxEVT_AUINOTEBOOK_ALLOW_DND,
            &AuiNotebook::OnAllowTabDrop, this);
        window->Bind(wxEVT_AUINOTEBOOK_DRAG_DONE,
            &AuiNotebook::OnDragDone, this);
        window->Bind(wxEVT_AUINOTEBOOK_TAB_MIDDLE_DOWN,
            &AuiNotebook::OnTabMiddleMouseDown, this);
        window->Bind(wxEVT_AUINOTEBOOK_TAB_MIDDLE_UP,
            &AuiNotebook::OnTabMiddleMouseUp, this);
        window->Bind(wxEVT_AUINOTEBOOK_TAB_RIGHT_DOWN,
            &AuiNotebook::OnTabRightMouseDown, this);
        window->Bind(wxEVT_AUINOTEBOOK_TAB_RIGHT_UP,
            &AuiNotebook::OnTabRightMouseUp, this);
        window->Bind(wxEVT_AUINOTEBOOK_BG_DCLICK,
            &AuiNotebook::OnBgDclickMouse, this);
        return window;
    }

    wxAuiNotebook* AuiNotebook::GetNotebook()
    {
        return dynamic_cast<wxAuiNotebook*>(GetWxWindow());
    }

    void AuiNotebook::SetArtProvider(void* art)
    {
    }

    void* AuiNotebook::GetArtProvider()
    {
        return nullptr;
    }

    void AuiNotebook::SetUniformBitmapSize(int width, int height)
    {
    }

    void AuiNotebook::SetTabCtrlHeight(int height)
    {
    }

    bool AuiNotebook::AddPage(void* page, const string& caption,
        bool select, ImageSet* bitmap)
    {
        return false;
    }

    bool AuiNotebook::InsertPage(uint64_t pageIdx, void* page, const string& caption,
        bool select, ImageSet* bitmap)
    {
        return false;
    }

    bool AuiNotebook::DeletePage(uint64_t page)
    {
        return false;
    }

    bool AuiNotebook::RemovePage(uint64_t page)
    {
        return false;
    }

    uint64_t AuiNotebook::GetPageCount()
    {
        return 0;
    }

    void* AuiNotebook::GetPage(uint64_t pageIdx)
    {
        return nullptr;
    }

    int AuiNotebook::FindPage(void* page)
    {
        return 0;
    }

    bool AuiNotebook::SetPageText(uint64_t page, const string& text)
    {
        return false;
    }

    string AuiNotebook::GetPageText(uint64_t pageIdx)
    {
        return wxStr(wxEmptyString);
    }

    bool AuiNotebook::SetPageToolTip(uint64_t page, const string& text)
    {
        return false;
    }

    string AuiNotebook::GetPageToolTip(uint64_t pageIdx)
    {
        return wxStr(wxEmptyString);
    }

    bool AuiNotebook::SetPageBitmap(uint64_t page, ImageSet* bitmap)
    {
        return false;
    }

    int AuiNotebook::SetSelection(uint64_t newPage)
    {
        return 0;
    }

    int AuiNotebook::GetSelection()
    {
        return 0;
    }

    void AuiNotebook::Split(uint64_t page, int direction)
    {
    }

    int AuiNotebook::GetTabCtrlHeight()
    {
        return 0;
    }

    int AuiNotebook::GetHeightForPageHeight(int pageHeight)
    {
        return 0;
    }

    bool AuiNotebook::ShowWindowMenu()
    {
        return false;
    }

    bool AuiNotebook::DeleteAllPages()
    {
        return false;
    }
}

File: Source/Alternet.UI/Controls/AuiManager/AuiNotebook.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Alternet.Drawing;

namespace Alternet.UI
{
    /// <summary>
    /// Notebook control, managing multiple windows with associated tabs.
    /// Used with <see cref="AuiManager"/>.
    /// </summary>
    public class AuiNotebook : Control
    {
        /// <summary>
        /// Occurs when a page is about to be closed.
        /// </summary>
        public event EventHandler? PageClose;

        /// <summary>
        /// Occurs when a page has been closed.
        /// </summary>
        public event EventHandler? PageClosed;

        /// <summary>
        /// Occurs when the page selection was changed.
        /// </summary>
        public event EventHandler? PageChanged;

        /// <summary>
        /// Occurs when the page selection is about to be changed.
        /// This event can be vetoed.
        /// </summary>
        public event EventHandler? PageChanging;

        /// <summary>
        /// Occurs when the window list button has been pressed.
        /// </summary>
        public event EventHandler? WindowListButton;

        /// <summary>
        /// Occurs when dragging is about to begin.
        /// </summary>
        public event EventHandler? BeginDrag;

        /// <summary>
        /// Occurs when dragging has ended.
        /// </summary>
        public event EventHandler? EndDrag;

        /// <summary>
        /// Occurs during a drag and drop operation.
        /// </summary>
        public event EventHandler? DragMotion;

        /// <summary>
        /// Occurs when application needs to query whether to allow a tab
        /// to be dropped. This event must be specially allowed.
        /// </summary>
        public event EventHandler? AllowTabDrop;

        /// <summary>
        /// Occurs to notify that the tab has been dragged.
        /// </summary>
        public event EventHandler? DragDone;

        /// <summary>
        /// Occurs when the middle mouse button is pressed on a tab.
        /// </summary>
        public event EventHandler? TabMiddleMouseDown;

        /// <summary>
        /// Occurs when the middle mouse button is released on a tab.
        /// </summary>
        public event EventHandler? TabMiddleMouseUp;

        /// <summary>
        /// Occurs when the right mouse button is pressed on a tab.
        /// </summary>
        public event EventHandler? TabRightMouseDown;

        /// <summary>
        /// Occurs when the right mouse button is released on a tab.
        /// </summary>
        public event EventHandler? TabRightMouseUp;

        /// <summary>
        /// Occurs when the mouse button is double clicked on the tabs
        /// background area.
        /// </summary>
        public event EventHandler? BgDclickMouse;

        public int EventSelection
        {
            get
            {
                return NativeControl.EventSelection;
            }
        }

        public int EventOldSelection
        {
            get
            {
                return NativeControl.EventOldSelection;
            }
        }

        /// <inheritdoc/>
        public override ControlId ControlKind => ControlId.AuiNotebook;

        internal new NativeAuiNotebookHandler Handler
        {
            get
            {
                CheckDisposed();
                return (NativeAuiNotebookHandler)base.Handler;
            }
        }

        internal Native.AuiNotebook NativeControl => Handler.NativeControl;

        /// <summary>
        /// Ensures that all tabs have the same height, even if some of them
        /// don't have bitmaps.
        /// </summary>
        /// <param name="width">Bitmap width for tabs without images.</param>
        /// <param name="height">Bitmap height for tabs without images.</param>
        /// <remarks>
        /// Passing (-1,-1) as size undoes the effect of a previous call to this
        /// function and instructs the control to use dynamic tab height.
        /// </remarks>
        public void SetUniformBitmapSize(int width, int height)
        {
            NativeControl.SetUniformBitmapSize(width, height);
        }

        /// <summary>
        /// Sets the tab height.
        /// </summary>
        /// <param name="height">New tab height value.</param>
        /// <remarks>
        /// By default, the tab control height is calculated by measuring the
        /// text height and bitmap sizes on the tab captions. Calling this
        /// method will override that calculation and set the tab control
        /// to the specified height parameter. A call to this method will
        /// override any call to <see cref="SetUniformBitmapSize"/>.
        /// </remarks>
        /// <remarks>
        /// Specifying -1 as the height will return the control to its
        /// default auto-sizing behaviour.
        /// </remarks>
        public void SetTabCtrlHeight(int height)
        {
            NativeControl.SetTabCtrlHeight(height);
        }

        /// <summary>
        /// Adds a new page.
        /// </summary>
        /// <param name="page">Controls for the new page.</param>
        /// <param name="caption">Text for the new page.</param>
        /// <param name="select">Specifies whether the page should be selected.</param>
        /// <param name="bitmap">image for the new page. Optional.</param>
        /// <returns><c>true</c> if successful, <c>false</c> otherwise.</returns>
        /// <remarks>
        /// If the <paramref name="select"/> parameter is true, calling
        /// this will generate a page change event.
        /// </remarks>
        /// <remarks>
        /// <see cref="InsertPage"/> is similar to <see cref="AddPage"/>, but allows
        /// the ability to specify the insert location.
        /// </remarks>
        public bool AddPage(
            Control page,
            string caption,
            bool select = true,
            ImageSet? bitmap = null)
        {
            return NativeControl.AddPage(
                page.WxWidget,
                caption,
                select,
                bitmap?.NativeImageSet);
        }

        /// <summary>
        /// Inserts a new page at the specified position.
        /// </summary>
        /// <param name="pageIdx">Position for the new page.</param>
        /// <param name="page">Controls for the new page.</param>
        /// <param name="caption">Text for the new page.</param>
        /// <param name="select">Specifies whether the page should be selected.</param>
        /// <param name="bitmap">image for the new page. Optional.</param>
        /// <returns><c>true</c> if successful, <c>false</c> otherwise.</returns>
        /// <remarks>
        /// If the <paramref name="select"/> parameter is true, calling
        /// this will generate a page change event.
        /// </remarks>
        /// <remarks>
        /// <see cref="InsertPage"/> is similar to <see cref="AddPage"/>, but allows
        /// the ability to specify the insert location.
        /// </remarks>
        public bool InsertPage(
            ulong pageIdx,
            Control page,
            string caption,
            bool select = true,
            ImageSet? bitmap = null)
        {
            return NativeControl.InsertPage(
                pageIdx,
                page.WxWidget,
                caption,
                select,
                bitmap?.NativeImageSet);
        }

        /// <summary>
        /// Deletes a page at the given index.
        /// </summary>
        /// <param name="page">Page index.</param>
        /// <returns><c>true</c> if operation was successfull,
        /// <c>false</c> otherwise.</returns>
        /// <remarks>
        /// Calling this method will generate a page change event.
        /// </remarks>
        public bool DeletePage(ulong page)
        {
            return NativeControl.DeletePage(page);
        }

        /// <summary>
        /// Removes a page, without deleting the attached control.
        /// </summary>
        /// <param name="page">Page index</param>
        /// <returns><c>true</c> if operation was successfull,
        /// <c>false</c> otherwise.</returns>
        public bool RemovePage(ulong page)
        {
            return NativeControl.RemovePage(page);
        }

        /// <summary>
        /// Returns the number of pages in the notebook.
        /// </summary>
        public ulong GetPageCount()
        {
            return NativeControl.GetPageCount();
        }

        /// <summary>
        /// Returns the page index for the specified control.
        /// </summary>
        /// <param name="page">Page index.</param>
        /// <remarks>
        /// If the window is not found in the notebook, -1 is returned.
        /// </remarks>
        public int FindPage(Control page)
        {
            return NativeControl.FindPage(page.WxWidget);
        }

        /// <summary>
        /// Sets the tab label for the page.
        /// </summary>
        /// <param name="page">Page index.</param>
        /// <param name="text">New tab label.</param>
        /// <returns></returns>
        public bool SetPageText(ulong page, string text)
        {
            return NativeControl.SetPageText(page, text);
        }

        /// <summary>
        /// Returns the tab label for the page.
        /// </summary>
        /// <param name="pageIdx">Page index.</param>
        public string GetPageText(ulong pageIdx)
        {
            return NativeControl.GetPageText(pageIdx);
        }

        /// <summary>
        /// Sets the tooltip displayed when hovering over the tab label of the page.
        /// </summary>
        /// <param name="page">Page index.</param>
        /// <param name="text">New tooltip value.</param>
        /// <returns><c>true</c> if tooltip was updated, <c>false</c> if it failed,
        /// e.g. because the page index is invalid.</returns>
        public bool SetPageToolTip(ulong page, string text)
        {
            return NativeControl.SetPageToolTip(page, text);
        }

        /// <summary>
        /// Returns the tooltip for the tab label of the page.
        /// </summary>
        /// <param name="pageIdx">Page index.</param>
        public string GetPageToolTip(ulong pageIdx)
        {
            return NativeControl.GetPageToolTip(pageIdx);
        }

        /// <summary>
        /// Sets the bitmap for the page.
        /// </summary>
        /// <param name="page">Page index.</param>
        /// <param name="bitmap">Bitmap for the page.</param>
        /// <remarks>
        /// To remove a bitmap from the tab caption, pass <c>null</c>.
        /// </remarks>
        public bool SetPageBitmap(ulong page, ImageSet? bitmap)
        {
            return NativeControl.SetPageBitmap(page, bitmap?.NativeImageSet);
        }

        /// <summary>
        /// Sets the page selection.
        /// </summary>
        /// <param name="newPage">New selected page index.</param>
        /// <remarks>
        /// Calling this method will generate a page change event.
        /// </remarks>
        public long SetSelection(ulong newPage)
        {
            return NativeControl.SetSelection(newPage);
        }

        /// <summary>
        /// Returns the currently selected page index.
        /// </summary>
        /// <returns></returns>
        public long GetSelection()
        {
            return NativeControl.GetSelection();
        }

        /// <summary>
        /// Performs a split operation programmatically.
        /// </summary>
        /// <param name="page">Indicates the page that will be split off.
        /// This page will also become the active page after the split.</param>
        /// <param name="direction">Specifies where the pane should go, it should
        /// be one of the following: Top, Bottom, Left, or Right.</param>
        /// <exception cref="ArgumentException"><paramref name="direction"/>
        /// is incorrect.</exception>
        public void Split(ulong page, GenericDirection direction)
        {
            if (direction == GenericDirection.Top || direction == GenericDirection.Bottom
                || direction == GenericDirection.Right ||
                direction == GenericDirection.Left)
                NativeControl.Split(page, (int)direction);
            else
                throw new ArgumentException(nameof(direction));
        }

        /// <summary>
        /// Returns the height of the tab control.
        /// </summary>
        public int GetTabCtrlHeight()
        {
            return NativeControl.GetTabCtrlHeight();
        }

        /// <summary>
        /// Gets the height of the notebook for a given page height.
        /// </summary>
        /// <param name="pageHeight">Page height.</param>
        /// <returns>The desired height of the notebook for the given
        /// page height.</returns>
        /// <remarks>
        /// Use this to fit the notebook to a given page size.
        /// </remarks>
        public int GetHeightForPageHeight(int pageHeight)
        {
            return NativeControl.GetHeightForPageHeight(pageHeight);
        }

        /// <summary>
        /// Shows the window menu for the active tab control associated with
        /// this notebook, and returns true if a selection was made.
        /// </summary>
        /// <returns></returns>
        public bool ShowWindowMenu()
        {
            return NativeControl.ShowWindowMenu();
        }

        /// <summary>
        /// Deletes all pages.
        /// </summary>
        /// <returns><c>true</c> if operation was performed successfully.</returns>
        public bool Clear()
        {
            return NativeControl.DeleteAllPages();
        }

        /// <summary>
        /// Changes the selection for the given page, returning the previous selection.
        /// </summary>
        /// <param name="newPage">New selected page index.</param>
        /// <remarks>
        /// This function behaves as <see cref="SetSelection"/> but does not
        /// generate the page changing events.
        /// </remarks>
        public long ChangeSelection(ulong newPage)
        {
            return NativeControl.ChangeSelection(newPage);
        }

        /// <summary>
        /// Selects next or previous page.
        /// </summary>
        /// <param name="forward">If <c>true</c> selects next page, if
        /// <c>false</c> selects previous page.</param>
        public void AdvanceSelection(bool forward = true)
        {
            NativeControl.AdvanceSelection(forward);
        }

        /// <summary>
        /// Sets the font for measuring tab labels.
        /// </summary>
        /// <param name="font">New font value.</param>
        public void SetMeasuringFont(Font? font)
        {
            if (font == null)
                font = Font.Default;
            NativeControl.SetMeasuringFont(font.NativeFont);
        }

        /// <summary>
        /// Sets the font for drawing unselected tab labels.
        /// </summary>
        /// <param name="font">New font value.</param>
        public void SetNormalFont(Font? font)
        {
            if (font == null)
                font = Font.Default;
            NativeControl.SetNormalFont(font.NativeFont);
        }

        /// <summary>
        /// Sets the font for drawing selected tab labels.
        /// </summary>
        /// <param name="font">New font value.</param>
        public void SetSelectedFont(Font? font)
        {
            if (font == null)
                font = Font.Default;
            NativeControl.SetSelectedFont(font.NativeFont);
        }

        /// <summary>
        /// Returns the page specified by the given index.
        /// </summary>
        /// <param name="pageIdx">Page index.</param>
        internal IntPtr GetPage(ulong pageIdx)
        {
            return NativeControl.GetPage(pageIdx);
        }

        /// <summary>
        /// Sets the art provider to be used by the notebook.
        /// </summary>
        /// <param name="art">New art provider.</param>
        internal void SetArtProvider(IntPtr art)
        {
            NativeControl.SetArtProvider(art);
        }

        internal void RaisePageClose(EventArgs e)
        {
            OnPageClose(e);
            PageClose?.Invoke(this, e);
        }

        internal void RaisePageClosed(EventArgs e)
        {
            OnPageClosed(e);
            PageClosed?.Invoke(this, e);
        }

        internal void RaisePageChanged(EventArgs e)
        {
            OnPageChanged(e);
            PageChanged?.Invoke(this, e);
        }

        internal void RaisePageChanging(EventArgs e)
        {
            OnPageChanging(e);
            PageChanging?.Invoke(this, e);
        }

        internal void RaiseWindowListButton(EventArgs e)
        {
            OnWindowListButton(e);
            WindowListButton?.Invoke(this, e);
        }

        internal void RaiseBeginDrag(EventArgs e)
        {
            OnBeginDrag(e);
            BeginDrag?.Invoke(this, e);
        }

        internal void RaiseEndDrag(EventArgs e)
        {
            OnEndDrag(e);
            EndDrag?.Invoke(this, e);
        }

        internal void RaiseDragMotion(EventArgs e)
        {
            OnDragMotion(e);
            DragMotion?.Invoke(this, e);
        }

        internal void RaiseAllowTabDrop(EventArgs e)
        {
            OnAllowTabDrop(e);
            AllowTabDrop?.Invoke(this, e);
        }

        internal void RaiseDragDone(EventArgs e)
        {
            OnDragDone(e);
            DragDone?.Invoke(this, e);
        }

        internal void RaiseTabMiddleMouseDown(EventArgs e)
        {
            OnTabMiddleMouseDown(e);
            TabMiddleMouseDown?.Invoke(this, e);
        }

        internal void RaiseTabMiddleMouseUp(EventArgs e)
        {
            OnTabMiddleMouseUp(e);
            TabMiddleMouseUp?.Invoke(this, e);
        }

        internal void RaiseTabRightMouseDown(EventArgs e)
        {
            OnTabRightMouseDown(e);
            TabRightMouseDown?.Invoke(this, e);
        }

        internal void RaiseTabRightMouseUp(EventArgs e)
        {
            OnTabRightMouseUp(e);
            TabRightMouseUp?.Invoke(this, e);
        }

        internal void RaiseBgDclickMouse(EventArgs e)
        {
            OnBgDclickMouse(e);
            BgDclickMouse?.Invoke(this, e);
        }

        /// <summary>
        /// Returns the associated art provider.
        /// </summary>
        internal IntPtr GetArtProvider()
        {
            return NativeControl.GetArtProvider();
        }

        /// <inheritdoc/>
        protected override ControlHandler CreateHandler()
        {
            return GetEffectiveControlHandlerHactory().CreateAuiNotebookHandler(this);
        }

        /// <summary>
        /// Called when a page is about to be closed.
        /// </summary>
        /// <param name="e">An <see cref="EventArgs"/> that contains
        /// the event data.</param>
        protected virtual void OnPageClose(EventArgs e)
        {
        }

        /// <summary>
        /// Occurs when a page has been closed.
        /// </summary>
        /// <param name="e">An <see cref="EventArgs"/> that contains
        /// the event data.</param>
        protected virtual void OnPageClosed(EventArgs e)
        {
        }

        /// <summary>
        /// Called when the page selection was changed.
        /// </summary>
        /// <param name="e">An <see cref="EventArgs"/> that contains
        /// the event data.</param>
        protected virtual void OnPageChanged(EventArgs e)
        {
        }

        /// <summary>
        /// Called when the page selection is about to be changed.
        /// This event can be vetoed.
        /// </summary>
        /// <param name="e">An <see cref="EventArgs"/> that contains
        /// the event data.</param>
        protected virtual void OnPageChanging(EventArgs e)
        {
        }

        /// <summary>
        /// Called when the window list button has been pressed.
        /// </summary>
        /// <param name="e">An <see cref="EventArgs"/> that contains
        /// the event data.</param>
        protected virtual void OnWindowListButton(EventArgs e)
        {
        }

        /// <summary>
        /// Called when dragging is about to begin.
        /// </summary>
        /// <param name="e">An <see cref="EventArgs"/> that contains
        /// the event data.</param>
        protected virtual void OnBeginDrag(EventArgs e)
        {
        }

        /// <summary>
        /// Called when dragging has ended.
        /// </summary>
        /// <param name="e">An <see cref="EventArgs"/> that contains
        /// the event data.</param>
        protected virtual void OnEndDrag(EventArgs e)
        {
        }

        /// <summary>
        /// Called during a drag and drop operation.
        /// </summary>
        /// <param name="e">An <see cref="EventArgs"/> that contains
        /// the event data.</param>
        protected virtual void OnDragMotion(EventArgs e)
        {
        }

        /// <summary>
        /// Occurs when application needs to query whether to allow a tab
        /// to be dropped. This event must be specially allowed.
        /// </summary>
        /// <param name="e">An <see cref="EventArgs"/> that contains
        /// the event data.</param>
        protected virtual void OnAllowTabDrop(EventArgs e)
        {
        }

        /// <summary>
        /// Called to notify that the tab has been dragged.
        /// </summary>
        /// <param name="e">An <see cref="EventArgs"/> that contains
        /// the event data.</param>
        protected virtual void OnDragDone(EventArgs e)
        {
        }

        /// <summary>
        /// Called when the middle mouse button is pressed on a tab.
        /// </summary>
        /// <param name="e">An <see cref="EventArgs"/> that contains
        /// the event data.</param>
        protected virtual void OnTabMiddleMouseDown(EventArgs e)
        {
        }

        /// <summary>
        /// Called when the middle mouse button is released on a tab.
        /// </summary>
        /// <param name="e">An <see cref="EventArgs"/> that contains
        /// the event data.</param>
        protected virtual void OnTabMiddleMouseUp(EventArgs e)
        {
        }

        /// <summary>
        /// Called when the right mouse button is pressed on a tab.
        /// </summary>
        /// <param name="e">An <see cref="EventArgs"/> that contains
        /// the event data.</param>
        protected virtual void OnTabRightMouseDown(EventArgs e)
        {
        }

        /// <summary>
        /// Called when the right mouse button is released on a tab.
        /// </summary>
        /// <param name="e">An <see cref="EventArgs"/> that contains
        /// the event data.</param>
        protected virtual void OnTabRightMouseUp(EventArgs e)
        {
        }

        /// <summary>
        /// Called when the mouse button is double clicked on the tabs
        /// background area.
        /// </summary>
        /// <param name="e">An <see cref="EventArgs"/> that contains
        /// the event data.</param>
        protected virtual void OnBgDclickMouse(EventArgs e)
        {
        }
    }
}